### DISCUSSION TITLE
merge several thousand images with arrayjoin

### DISCUSSION THREAD

**Discussion by heman1-test** [libvips/libvips#3755] [Q&A]
Hi,

how is it possible in Windows to join several thousand images with arrayjoin method?

Writing all filenames to cammand line at once is not possible

this is working limited to a few files only
**vips arrayjoin "00.jpg 01.jpg 02.jpg 03.jpg" x.jpg --across=2**

this is not working  but needed for several thousand files
**vips arrayjoin "*.jpg" x.jpg --across=2**

Any ideas how to handle that maybe with a for loop?

Thank You

**Comment by jcupitt** [comment]
You can do it with a `$()` expansion if you use bash on windows, eg.:

```
jcupi@DESKTOP-HGI6HBR MINGW64 /f/Pictures/test
$ /f/vips-dev-8.14/bin/vips.exe crop ../st-francis.jpg ../tile.jpg 0 0 500 500

jcupi@DESKTOP-HGI6HBR MINGW64 /f/Pictures/test
$ for i in {1..10000}; do cp ../tile.jpg $i.jpg; done

jcupi@DESKTOP-HGI6HBR MINGW64 /f/Pictures/test
$ time /f/vips-dev-8.14/bin/vips.exe arrayjoin "$(echo tile_*.jpg)" x.tif[tile,compression=jpeg] --across 100
bash: /f/vips-dev-8.14/bin/vips.exe: Argument list too long
```

But (at least with 10,000 jpgs) it'll fail with argument list too long. I think you'll need to use something like python.

**Comment by heman1-test** [comment]
Thank you John but then  I will also fail with your proposal. I have 97 x 88 (8536) files to handle with. Is there any way of chunking?

1. process the first row
2. save that workerfile
3. process 2nd row
4. save that dummy file 
5. join that 2 rows to new worker file and proceed with next row until end

**Comment by jcupitt** [comment]
Sure, you can assemble in sections. I would use tiled jpeg TIFF for the intermediates. Of course, it'll be slower and you'll need a lot of disc space.

Is python / C# / ruby / php / etc. not possible?

**Comment by jcupitt** [comment]
... I tried some other sizes. 70x70 (5,000) jpgs failed with arg list too long, but 50x50 (2,500 jpgs) worked fine. So the limit is somewhere in that range.

**Comment by heman1-test** [comment]
Thanks John for testing it.
Is it possible to implement a kind of input file list read from a text file? You can put in all files in needed order in row by row and can then read that textfile as input. This would save a lot of extra effort as well.

**Comment by jcupitt** [comment]
There's no built in way of doing that, no.

I'd either assemble in sections, or use python.

In python, it'd be:

```python
import sys
import glob
import pyvips

input_directory = sys.argv[1]
output_file = sys.argv[2]
tiles_across = int(sys.argv[3])

images = [pyvips.Image.new_from_file(filename, access="sequential")
          for filename in glob.glob(f"{input_directory}/*.jpg")]

joined = pyvips.Image.arrayjoin(images, across=tiles_across)

joined.write_to_file(output_file)
```

**Comment by heman1-test** [comment]
Thank you!
